indico/modules/events/ical.py

```diff
@@ -12,72 +12,135 @@
 
 from indico.core import signals
 from indico.core.config import config
+from indico.core.db.sqlalchemy.protection import ProtectionMode
 from indico.util.date_time import now_utc
 from indico.util.signals import values_from_signal
 
 
-def event_to_ical(event, user=None):
+def generate_basic_component(entity, uid=None, url=None):
+    """Generates an icalendar component with basic common properties.
+
+    :param entity: Event/session/contribution where properties come from
+    :param uid: UID for the component
+    :param url: URL for the component (defaults to `entity.external_url`)
+
+    :returns: icalendar event with basic properties
+    """
+
+    component = Event()
+
+    component.add('dtstamp', now_utc(False))
+    component.add('dtstart', entity.start_dt)
+    component.add('dtend', entity.end_dt)
+    component.add('summary', entity.title)
+
+    if uid:
+        component.add('uid', uid)
+
+    if url:
+        component.add('url', url)
+    elif hasattr(entity, 'external_url'):
+        component.add('url', entity.external_url)
+
+    location = (f'{entity.room_name} ({entity.venue_name})'
+                if entity.venue_name and entity.room_name
+                else (entity.venue_name or entity.room_name))
+    if location:
+        component.add('location', location)
+
+    # speakers are located in a different variable in contributions
+    speaker_list = None
+    if hasattr(entity, 'person_links'):
+        speaker_list = entity.person_links
+    elif hasattr(entity, 'speakers'):
+        speaker_list = entity.speakers
```

Comment:
This won't work as you expect: contributions also have `person_links` (authors, speakers, etc.) so you need to check for speakers first.

If all entities used here have at least `person_links` you can just use this:

```python
speaker_list = getattr(entity, 'speakers', entity.person_links)
```

Otherwise just reverse the order of your if/elif.

Reply:
Actually, the original comment is incorrect. It is sessions which don't have `person_links`, but they do not have `speakers` either, so I think removing the elif could also work.